pkg/assembler/backends/testing/pkg.go

```diff
@@ -17,206 +17,405 @@ package testing
-	for i, _ := range qualifiers {
-		if i%2 == 0 {
-			qs = append(qs, &model.PackageQualifier{
-				Key:   qualifiers[i],
-				Value: qualifiers[i+1],
-			})
+	node, ok := index[id]
+	if !ok {
+		return nil, gqlerror.Errorf("ID does not match existing node")
+	}
+
+	pvl := []*model.PackageVersion{}
+	if versionNode, ok := node.(*pkgVersionNode); ok {
+		pv := model.PackageVersion{
+			// IDs are generated as string even though we ask for integers
+			// See https://github.com/99designs/gqlgen/issues/2561
+			ID:         fmt.Sprintf("%d", versionNode.id),
+			Version:    versionNode.version,
+			Subpath:    versionNode.subpath,
+			Qualifiers: getCollectedPackageQualifiers(versionNode.qualifiers),
+		}
+		if filter != nil && noMatch(filter.Version, pv.Version) {
+			return nil, nil
+		}
+		if filter != nil && noMatch(filter.Subpath, pv.Subpath) {
+			return nil, nil
+		}
+		if filter != nil && noMatchQualifiers(filter, versionNode.qualifiers) {
+			return nil, nil
 		}
+		pvl = append(pvl, &pv)
+		node = index[versionNode.parent]
 	}
-	return qs
-}
 
-// Query Package
+	pnl := []*model.PackageName{}
+	if versionStruct, ok := node.(*pkgVersionStruct); ok {
+		pn := model.PackageName{
+			// IDs are generated as string even though we ask for integers
+			// See https://github.com/99designs/gqlgen/issues/2561
+			ID:       fmt.Sprintf("%d", versionStruct.id),
+			Name:     versionStruct.name,
+			Versions: pvl,
+		}
+		if filter != nil && noMatch(filter.Name, pn.Name) {
+			return nil, nil
+		}
+		pnl = append(pnl, &pn)
+		node = index[versionStruct.parent]
+	}
 
-func (c *demoClient) Packages(ctx context.Context, pkgSpec *model.PkgSpec) ([]*model.Package, error) {
-	var packages []*model.Package
-	for _, p := range c.packages {
-		if pkgSpec.Type == nil || p.Type == *pkgSpec.Type {
-			newPkg := filterPackageNamespace(p, pkgSpec)
-			if newPkg != nil {
-				packages = append(packages, newPkg)
-			}
+	pnsl := []*model.PackageNamespace{}
+	if nameStruct, ok := node.(*pkgNameStruct); ok {
+		pns := model.PackageNamespace{
+			// IDs are generated as string even though we ask for integers
+			// See https://github.com/99designs/gqlgen/issues/2561
+			ID:        fmt.Sprintf("%d", nameStruct.id),
+			Namespace: nameStruct.namespace,
+			Names:     pnl,
+		}
+		if filter != nil && noMatch(filter.Namespace, pns.Namespace) {
+			return nil, nil
 		}
+		pnsl = append(pnsl, &pns)
+		node = index[nameStruct.parent]
+	}
+
+	namespaceStruct, ok := node.(*pkgNamespaceStruct)
+	if !ok {
+		return nil, gqlerror.Errorf("ID does not match expected node type")
+	}
+	p := model.Package{
+		// IDs are generated as string even though we ask for integers
+		// See https://github.com/99designs/gqlgen/issues/2561
+		ID:         fmt.Sprintf("%d", namespaceStruct.id),
+		Type:       namespaceStruct.typeKey,
+		Namespaces: pnsl,
 	}
-	return packages, nil
+	if filter != nil && noMatch(filter.Type, p.Type) {
+		return nil, nil
+	}
+	return &p, nil
 }
 
+// Query Package
+
```

Comment:
This should be on top of `Packages`.

I think we can also delete `filterPackageNamespace`

Reply:
we need `filterPackageNamespace` for the time being until we replace other instances where its being used. It will be removed during this cleanup

Reply:
added a TODO to cleanup